### PR TITLE
Added Numba RV conversion

### PR DESCRIPTION
**Pull request by kc611**
From what I gather :

 `_helperlib` stores a seperate state `ptr` value for Numba library. So during `RandomState` typification, we can sync this internal Numba state with Numpy's `RandomVariable` state. (However I'm not too sure that this will work every time.).

As for RandomVariable, it would be as simple as jitting the original Numpy function, as Numba has direct support for most of Numpy's `.random` methods. However Numba's implementations don't support `size` arguments. A simple way to overcome this would be looping over an empty array. However that's inefficient and I'm sure there are other better ways to do that. (Suggestions?)

### CODE REVIEW

**Comment by twiecki** [comment]
Thanks for taking this on so quickly.
1. We should open an issue for numba to support `size`.
2. For looping: As this gets JITed by `numba` do we really expect this to be slow?
3. Can we use something like `@vectorize` or `@guvectorize` (https://numba.pydata.org/numba-doc/dev/user/vectorize.html)?

File: aesara/link/numba/dispatch.py [hunk 1]
```diff
@@ -1804,3 +1814,18 @@ def batched_dot(x, y):
...
 # optimizations are apparently already performed by Numba
+
+
+@numba_funcify.register(RandomVariable)
+def numba_funcify_RandomVariable(op, node, **kwargs):
+    name = op.name
+    random_func = getattr(np.random, name)
+
+    @numba.njit
+    def random_variable(rng, size, dtype, *args):
```
[the remark this answers is not in the document]
Comment (twiecki) [reply aesara/link/numba/dispatch.py:1825]: Got it, makes sense.

File: aesara/link/numba/linker.py [hunk 2]
```diff
@@ -12,15 +12,26 @@ def fgraph_convert(self, fgraph, **kwargs):
...
-            # TODO:When RandomVariable conversion is implemented
-            # do RandomState typification over here.
+            if isinstance(sinput[0], RandomState):
+                new_value = numba_typify(
+                    sinput[0], dtype=getattr(sinput[0], "dtype", None)
+                )
+                # We need to remove the reference-based connection to the
+                # original `RandomState`/shared variable's storage, because
+                # subsequent attempts to use the same shared variable within
+                # other non-JAXified graphs will have problems.
```
Comment (brandonwillard) [inline aesara/link/numba/linker.py:33]: This line mentions JAX.

File: aesara/link/numba/linker.py [hunk 3]
```diff
@@ -12,15 +12,26 @@ def fgraph_convert(self, fgraph, **kwargs):
...
+                sinput = [new_value]
```
Comment (brandonwillard) [inline aesara/link/numba/linker.py:34]: We should consider changing `numba_typify` so that it returns the "storage" list instead of just the contents of said list.  It seems like that would allow us to remove the explicit `RandomState` condition in this module.

File: aesara/link/numba/dispatch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (brandonwillard) [reply aesara/link/numba/dispatch.py:1825]: Writing a manual loop here is find&mdash;if necessary&mdash;because Numba can produce very performant loops.  My main concern is the overhead implied by a large number of independent calls to the underlying sampler code, instead of one call that handles the `size` parameter.  I still need to look at the Numba implementation to get a sense of that overhead, if any.

Regardless, a loop will work for now, and, to handle arbitrary dimensions in such a loop, we can just use `np.ndindex`.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/link/numba/dispatch.py [hunk 4]
```diff
@@ -283,9 +285,22 @@ def create_tuple_string(x):
 
 @singledispatch
 def numba_typify(data, dtype=None, **kwargs):
+    try:
+        for i, item in enumerate(data):
+            data[i] = numba_typify(item)
```
Comment (kc611) [inline aesara/link/numba/dispatch.py:290]: @brandonwillard Is this what you meant by changing `numba_typify` so that it returns the "storage" list ?

Comment (brandonwillard) [reply aesara/link/numba/dispatch.py:290]: Yeah, just an implementation that takes/uses and returns a "storage" list, so that we can keep the `RandomState`-specific requirements in `numba_typify_RandomState` (i.e. recreate the "storage" list in _there_).

This particular change makes the primary `data` argument a `list`, though, which means we can't recreate a "storage" list in the dispatch specializations when/if that's needed.  Instead, we can change the first/`data` argument into a `storage` argument that contains the single-element "storage" list (i.e. `state` in `numba_typify_RandomState` will be a list containing the `RandomState` object) and make `numba_typify` return the same list.

Comment (kc611) [reply aesara/link/numba/dispatch.py:290]: I don't exactly understand how this will be implemented though.

If you don't mind, can you go ahead and implement what you have in mind in this PR. (Cause the current implementation in this PR will have quite bit of overhead because of looping over each data member.)

**Comment by twiecki** [comment]
@kc611 This is shaping up nicely, what's missing?

**Comment by kc611** [comment]
This pretty much is the core implementation for Numba RV. Anything we add now will probably be optimization the existing code.

File: aesara/link/numba/dispatch.py [hunk 5]
```diff
@@ -1804,3 +1814,24 @@ def batched_dot(x, y):
...
+def numba_funcify_RandomVariable(op, node, **kwargs):
+    name = op.name
+    random_func = getattr(np.random, name)
+
+    tuple_type = numba.types.UniTuple(numba.types.intp, node.inputs[1].data.size)
+
+    @numba.njit
+    def random_variable(rng, size, dtype, *args):
+        with numba.objmode(size=tuple_type):
+            size = tuple(size)
```
Comment (brandonwillard) [inline aesara/link/numba/dispatch.py:1829]: We should be able to avoid this costly jump to object mode by using `numba.np.unsafe.ndarray.to_fixed_tuple` or `create_tuple_creator` and `tuple_setitem`.  See [`numba_funcify_Reshape`](https://github.com/pymc-devs/aesara/blob/abf6026c9c2bad19216275e127e3043f30d93360/aesara/link/numba/dispatch.py#L965) for an example of the latter.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/link/numba/dispatch.py (hunk 5 above)
Comment (brandonwillard) [reply aesara/link/numba/dispatch.py:1829]: I just tried changing `numba_funcify_Reshape` so that it uses `to_fixed_tuple`, and that approach works well&mdash;while also being a lot more succinct.

Comment (brandonwillard) [reply aesara/link/numba/dispatch.py:1829]: [Here's a PR](https://github.com/pymc-devs/aesara/pull/413) with those changes.

File: aesara/link/numba/dispatch.py [hunk 6]
```diff
@@ -1796,3 +1806,23 @@ def batched_dot(x, y):
...
 # they're only used to optimize basic `Dot` nodes, and those GEMV and GEMM
 # optimizations are apparently already performed by Numba
+
+
+@numba_funcify.register(RandomVariable)
+def numba_funcify_RandomVariable(op, node, **kwargs):
+    name = op.name
+    random_func = getattr(np.random, name)
+
+    tuple_size = node.inputs[1].data.size
```
Comment (brandonwillard) [inline aesara/link/numba/dispatch.py:1816]:
```suggestion
    tuple_size = node.inputs[1].ndim
```

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by brandonwillard** [comment]
I've added support for distribution parameters, but the current implementation still lacks some critical functionality.  Namely, it does not support distributions that aren't simply unadulterated calls to `numpy.random`.

This means that we need to restrict this Numba implementation to only the `RandomVariable`s that fit the aforementioned criteria.

Also, I noticed that the `Op.perform`-based approach won't work with `RandomVariable`s, because Numba cannot lower the `RandomState` returned by those `Op`s.  They're returned from Numba's object mode, so, no, object mode can't help us here.  

In other words, **we need to add support for `RandomState` types in Numba, or circumvent `RandomState`s altogether in `RandomVariable.perform` or our use of it (in this case)**.

The former is the prefered approach, of course.

File: aesara/tensor/random/basic.py [hunk 7]
```diff
@@ -81,6 +81,19 @@ def __call__(self, loc=0.0, scale=1.0, size=None, **kwargs):
...
+    @classmethod
+    def numba_fn(cls, np_random_fn_name, np_input_names, np_global_env):
+        np_global_env["abs_val"] = abs
+        np_global_env["standard_norm"] = np.random.standard_normal
+
+        src = f"""
+@numba_vectorize
+def {np_random_fn_name}({np_input_names}):
+    return abs_val(standard_norm({np_input_names}))
+"""
```
Comment (kc611) [inline aesara/tensor/random/basic.py:93]: This is something I tried (without much success) as an attempt to support custom implemented distributions. Somehow the output from this function gets classified as a `pyobject` rather than the `numba` type of the output value. (Even though both of the functions used inside are supported by Numba individually.)

Note: This particular implementation is taken from `scipy.stats` which does something pretty close to this.

https://github.com/scipy/scipy/blob/36af73c98f71f40bc69efbeb4501b1f2b9c3d00c/scipy/stats/_continuous_distns.py#L3504

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/link/numba/dispatch.py [hunk 8]
```diff
@@ -1816,27 +1816,50 @@ def numba_funcify_RandomVariable(op, node, **kwargs):
...
-    """
-    np_random_fn = compile_function_src(
-        np_random_fn_src, np_random_fn_name, np_global_env
-    )
+"""
+        np_random_fn = compile_function_src(
+            np_random_fn_src, np_random_fn_name, np_global_env
+        )
+
+    elif hasattr(op, "numba_fn"):
```
[the remark this answers is not in the document]
Comment (brandonwillard) [reply aesara/link/numba/dispatch.py:1840]: Oh, also, if you we want to simplify the process of creating dispatch functions and/or keep them closer/within to their class definitions, we can consider doing something like we did for `Distribution` (see [here](https://github.com/pymc-devs/pymc3/blob/v4/pymc3/distributions/distribution.py#L99)).

[the remark this answers is not in the document]
Comment (kc611) [reply aesara/link/numba/dispatch.py:1840]: > we can just create new numba_funcify_* functions for each custom implementation instead.

Sure, we can do this too. But I recommend we at-least put the implementation of specific `RandomVariable`s in a different file cause it'll become more and more cluttered as we start to support more and more custom RV's. 

The exact implementation isn't the issue here though, I was (and still am) concerned  with why it isn't working. Or rather what am I missing here?

[the remark this answers is not in the document]
Comment (brandonwillard) [reply aesara/link/numba/dispatch.py:1840]: > But I recommend we at-least put the implementation of specific `RandomVariable`s in a different file cause it'll become more and more cluttered as we start to support more and more custom RV's.

Definitely; we should've already split `*.numba.dispatch` into separate modules.

**Comment by kc611** [comment]
> This means that we need to restrict this Numba implementation to only the RandomVariables that fit the aforementioned criteria.

What's the preferred approach we should be taking here or supporting the other `RandomVariables`?. I tried that experimentation with `halfnormal` and deriving it from `standard_normal`. But was unsuccessful so far (without falling back to object mode).  There are always inconsistencies between the exact values.

Or is that something we should be adding later on, and hold on till we can add `RandomState` types in Numba?

**Comment by brandonwillard** [comment]
> What's the preferred approach we should be taking here or supporting the other `RandomVariables`

For now, we can create distinct dispatch functions for all the `RandomVariable`s that are based on `np.random` samplers, instead of trying to make a single general Numba implementation.  

Really, we can just add a bunch of `@numba_funcify.register(...)`s onto the one function that handles the simple case of directly calling `np.random.*` and create separate implementations for the rest.

Otherwise, we need object mode to work for the rest, and I think your `numba_typify_RandomState` changes do that.  We just need some good tests for those cases.

The most important aspect of this Numba work is that we can cover all `Op`s by using their Python implementations when necessary, because this is what allows us to ultimately simplify and remove the old Theano backend.

File: aesara/link/numba/dispatch.py [hunk 9]
```diff
@@ -1865,3 +1877,94 @@ def batched_dot(x, y):
...
+    return np_random_func({np_input_names})
+    """
+    np_random_fn = compile_function_src(
+        np_random_fn_src, np_random_fn_name, np_global_env
+    )
+
+    return make_numba_random_fn(node, np_random_fn, np_input_names, np_random_fn_name)
+
+
+@numba_funcify.register(HalfNormalRV)
```
Comment (kc611) [inline aesara/link/numba/dispatch.py:1951]: For now, we can get away with this way of implementing non numpy RV's but it'll be nice to have a 'object mode' to fall back to if this fails. (Since, this particular implementation assumes we can convert any RV from `scipy.stats` to it's corresponding Numpy/Numba version, but that might not be always the case)

Comment (brandonwillard) [reply aesara/link/numba/dispatch.py]: Exactly!

**Review by brandonwillard — COMMENTED** [review]
These additions are looking pretty good.

I rebased, added the new unique-naming function (it also makes sure unique names are given to each function argument, even when some of the inputs are the same nodes), and enumerated the dispatches for all the Numba-supported `RandomVariable`s.  

We need to add some tests for&mdash;at least&mdash;the more "non-standard" distributions, like `RandIntRV`, `ChoiceRV` and `PermutationRV`.  We should probably have at least one test for every distribution, though.

Also, I think we've hit a file size extreme, and we should split apart `aesara.link.numba.dispatch` and the corresponding tests into separate modules.

**Comment by kc611** [comment]
It was probably a good idea to add separate tests for each RV's (we've got failures in some that'll need fixing).

Other than that I've made a separate files for dispatches in `aesara.link.numba.modules` but I cant find a way to connect them to the original `numba_funcify` in `aesara.link.numba.dispatch` (pre-commit basically does not allow module imports at end of file). Any suggestions how to do that?

**Review by brandonwillard — CHANGES_REQUESTED** [review]
I've updated the new sub-package/module layout and set the tests to `xfail` for now.

I could get some of the tests to work by emulating the `numba.tests.test_random.TestRandom._follow_numpy` approach; however, some Numba samplers apparently use CPython's random sampling.  One example is the `np.random.beta` sampler.

In these cases, I don't know if we can exactly compare NumPy/Aesara results with Numba.  We could instead compare the Aesara-generated Numba results with Numba itself.

File: tests/link/numba/utils.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+            return numpy_fn(*inputs)
+
+    else:
+        size = tuple(size.data.astype(np.int))
+
+        @numba.njit
+        def numba_fn():
+            data = np.empty(size)
+            for i in np.ndindex(size):
+                data[i] = numpy_fn(*inputs)
```
Comment (kc611) [inline tests/link/numba/utils.py:227]: Just a minor question. Over here, for distributions that accept arrays as inputs. (For instance `beta`) this doesn't accept anything other than a `float` value. If so then how is it that we can pass arrays (as `dvector`s) in Aesara graphs with Numba ? (like done in `test_random.test_RandomVariable`)

**Comment by brandonwillard** [comment]
FYI: The current failures are due to `pytest` and its [`import-mode` setting](https://docs.pytest.org/en/6.2.x/pythonpath.html).  The `import-mode` needed to be changed due to the presence of new sub-packages named `numba`.  

Basically, the way `pytest` takes over the Python import processes creates issues for `import numba` statements under a subpackage with the name `numba`.  Changing `import-mode` fixes that, but introduces a new issue for `pickle`.  See [this issue](https://github.com/pytest-dev/pytest/issues/7245) for more information.

File: tests/link/numba/utils.py (hunk 10 above)
Comment (brandonwillard) [reply tests/link/numba/utils.py:227]: We're `numba.vectorize`ing these `np.random` functions in Aesara.  That's what makes them work with array inputs.

We can do the same for testing.

File: tests/link/numba/utils.py [hunk 11]
```diff
@@ -211,11 +211,15 @@ def compare_rv_with_numba_output(fgraph, inputs, rng, size):
 
     inputs = tuple([i.tolist() for i in inputs])
 
+    @numba.vectorize
+    def numba_rng_fn(a):
+        return numpy_fn(*inputs)
```
Comment (kc611) [inline tests/link/numba/utils.py:216]: I tried doing a `numba.vectorize` but it still doesn't work with array inputs. There's probably something else I'm missing here.

Also a bit hackish but it seems functions compiled with `numba.vectorize` (without the decorator defining types, probably), they don't accept functions without any arguments. So I passed a single dummy variable(that `a`). Is this an issue with Numba or is it supposed to work this way.

Comment (brandonwillard) [reply tests/link/numba/utils.py:216]: The body of the function is supposed to represent the _scalar_ operation that's broadcast across arrays (i.e. "vectorized").

Comment (kc611) [reply tests/link/numba/utils.py]: > Varargs (i.e. *inputs) might be an issue for numba.vectorize, though.

Yeah, that's a problem. We can just use the `make_numba_random_fn` in this case, If our goal at the end is just comparing the inputs with and without the Numba function being in the graph.

**Comment by brandonwillard** [comment]
Ugh, I should've used something like `git merge` to split apart those files, because now these upstream changes need to be added manually.

File: aesara/link/numba/dispatch.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (brandonwillard) [reply aesara/link/numba/dispatch.py]: We'll do this in a separate PR; I don't want to hold this up any longer.
